source/common/router/rds_impl.cc

```diff
@@ -194,11 +205,30 @@ Router::ConfigConstSharedPtr RdsRouteConfigProviderImpl::config() {
   return tls_->getTyped<ThreadLocalConfig>().config_;
 }
 
+bool RdsRouteConfigProviderImpl::requestVirtualHostsUpdate(const std::string& for_domain,
+                                                           std::function<void()> cb) {
+  if (!config()->usesVhds()) {
+    return false;
+  }
+  factory_context_.dispatcher().post(
+      [this, for_domain]() -> void { subscription_->ondemandUpdate({for_domain}); });
+  config_update_callbacks_->getTyped<ThreadLocalCallbacks>().callbacks_.push(cb);
+
+  return true;
+}
+
 void RdsRouteConfigProviderImpl::onConfigUpdate() {
   ConfigConstSharedPtr new_config(
       new ConfigImpl(config_update_info_->routeConfiguration(), factory_context_, false));
-  tls_->runOnAllThreads(
-      [this, new_config]() -> void { tls_->getTyped<ThreadLocalConfig>().config_ = new_config; });
+  tls_->runOnAllThreads([this, new_config]() -> void {
+    tls_->getTyped<ThreadLocalConfig>().config_ = new_config;
+    auto callbacks = config_update_callbacks_->getTyped<ThreadLocalCallbacks>().callbacks_;
+    if (!callbacks.empty()) {
```

Comment:
Should this be a `while`? What if there are multiple updates requested before a response is received? Do you have tests for this?